lib/vsc/filesystem/quota/itools.py

```diff
@@ -0,0 +1,184 @@
+#
+# You should have received a copy of the GNU Library General Public License
+# along with vsc-filesystems-quota. If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+Helper functions for all things quota related.
+
+@author: Andy Georges (Ghent University)
+@author: Ward Poelmans (Vrije Universiteit Brussel)
+"""
+
+import gzip
+import json
+import logging
+import os
+import socket
+import time
+
+from collections import namedtuple
+
+from vsc.utils.script_tools import CLI
+
+from vsc.filesystem.gpfs import GpfsOperations
+from vsc.filesystem.lustre import LustreOperations
+from vsc.config.base import GENT, INSTITUTE_ADMIN_EMAIL
+from vsc.utils.mail import VscMail
+
+NAGIOS_CHECK_INTERVAL_THRESHOLD = (6 * 60 + 5) * 60  # 365 minutes -- little over 6 hours.
+INODE_LOG_ZIP_PATH = '/var/log/quota/inode-zips'
+INODE_STORE_LOG_CRITICAL = 1
+
+InodeCritical = namedtuple("InodeCritical", ['used', 'allocated', 'maxinodes'])
+
+CRITICAL_INODE_COUNT_MESSAGE = """
+Dear HPC admins,
+
+The following filesets will be running out of inodes soon (or may already have run out).
+
+%(fileset_info)s
+
+Kind regards,
+Your friendly inode-watching script
+"""
+
+
+def process_inodes_information(filesets, quota, threshold=0.9, storage='gpfs'):
+    """
+    Determines which filesets have reached a critical inode limit.
+
+    For this it uses the inode quota information passed in the quota argument and compares this with the maximum number
+    of inodes that can be allocated for the given fileset. The default threshold is placed at 90%.
+
+    @returns: dict with (filesetname, InodeCritical) key-value pairs
+    """
+    critical_filesets = dict()
+
+    for (fs_key, fs_info) in filesets.items():
+        allocated = int(fs_info['allocInodes']) if storage == 'gpfs' else 0
+        maxinodes = int(fs_info['maxInodes']) if storage == 'gpfs' else int(quota[fs_key][0].filesLimit)
+        used = int(quota[fs_key][0].filesUsage)
+
+        if maxinodes > 0 and used > threshold * maxinodes:
+            critical_filesets[fs_info['filesetName']] = InodeCritical(used=used, allocated=allocated,
+                                                                      maxinodes=maxinodes)
+
+    return critical_filesets
+
+
+class InodeLog(CLI):
+
+
+    # Note: debug option is provided by generaloption
+    # Note: other settings, e.g., ofr each cluster will be obtained from the configuration file
+    CLI_OPTIONS = {
+        'nagios-check-interval-threshold': NAGIOS_CHECK_INTERVAL_THRESHOLD,
+        'location': ('path to store the gzipped files', None, 'store', INODE_LOG_ZIP_PATH),
+        'backend': ('Storage backend', None, 'store', 'gpfs'),
+        'host_institute': ('Name of the institute where this script is being run', str, 'store', GENT),
+        'mailconfig': ("Full configuration for the mail sender", None, "store", None),
+    }
+
+    def mail_admins(self, critical_filesets, dry_run=True, host_institute=GENT):
```

Comment:
should the dry_run not be a CLI option as well? and then give none of these parameters a value?

Reply:
Not going to toch that, it works as is since the CLI_OPTIONS are merged with the default ones,